Data_Engine/Query/SubGraph.cs

```diff
@@ -0,0 +1,63 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Data.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
+using System;
+
+namespace BH.Engine.Data
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+        [Description("Extracts a sub graph from a graph")]
+        [Input("graph", "The graph to extract the sub graph from")]
+        [Input("rootNode", "The graph node that is the root of the sub graph")]
+        public static Graph<T> SubGraph<T>(this Graph<T> graph, GraphNode<T> rootNode)
+        {
+            Graph<T> subgraph = new Graph<T>();
+            subgraph.Nodes.Add(rootNode);
+            if (!graph.Nodes.Contains(rootNode))
+                throw new ArgumentException("rootnode provided cannot be found in the original graph. Ensure the node is from the original graph");
+            List<GraphLink<T>> links = graph.Links.FindAll(x => x.EndNode.Value.Equals(rootNode.Value));
```

Comment:
It would be safer to use the same technique you have used for `DepthDictionary` to avoid all the issues mentioned above about direction of the links.

Reply:
I'm looking at some background here. https://mathworld.wolfram.com/InducedSubgraph.html
According to that source the naming I have used is too vague. Looks like induced subgraphs graphs are really either edge induced or link induced. Neither of these applies to what I have needed.
Here what I need for the particular use is all "downstream" nodes in a decision graph. So maybe this one not generic enough to be here. Suggesting this is removed and located in the specific toolkit. What do you think?

Reply:
Or rename to be more specific?

Reply:
As per the other comment _downstream_ makes sense only for directed graphs. `SubGraph` makes sense to me if the recipe takes a list of nodes, or edges, as you say, and the methods figures out, respectively, either the connections, or the nodes.

I really think we should start splitting the two classes,

Reply:
> I really think we should start splitting the two classes,

@epignatelli 
Is this a suggestion we should have classes for `DirectedGraph` and `UndirectedGraph`?

Reply:
Not sure I like that idea too much if we can avoid it. The only difference between those two graphs is inside the links themselves. So we could just have one type of graph and two types of links (not necessarily two classes, it can just be a Boolean poverty of the link)

Reply:
Would it be simpler if there is a property like `IsDirected` on the `Graph` rather than on every link? 

Pleased the discussion is rolling a little 😃 now.

I think this is in need of raising a new issue and beyond the scope of the present PR - I have removed the SubGraph method from this PR so we can discuss further.

Reply:
> I think this is in need of raising a new issue and beyond the scope of the present PR

Agreed

Reply:
Yeah, I am fine with having only one class, as long as any manipulation of `Graph` makes sense.